From ebbd94b7107d4e8c90d72bd2cbca43127eab7051 Mon Sep 17 00:00:00 2001
From: Konstantin Porotchkin <kostap@marvell.com>
Date: Sun, 16 Sep 2018 10:47:53 +0300
Subject: [PATCH 0707/1051] dts: mvebu: Initial support for a3900 platform

Add initial support for A3900 platform and armada-3900-db-vd boards.
Working interfaces:
- PCIe0 and PCIe1
- NAND (8-bit ECC)
- I2C
- SPI
Not working:
- USB3 VBUS regulator
- PPv2 network ports (require PHY driver porting)

Change-Id: I63448c18dbbac31f4a1582652592a0f07b616880
Signed-off-by: Konstantin Porotchkin <kostap@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/60042
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Igal Liberman <igall@marvell.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06.
Just some minor context mods in order to port to wrlinux]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 arch/arm64/boot/dts/marvell/Makefile          |   2 +
 .../boot/dts/marvell/armada-3900-db-vd-A.dts  | 102 ++++++++++++
 .../boot/dts/marvell/armada-3900-db-vd-B.dts  | 103 ++++++++++++
 .../boot/dts/marvell/armada-3900-db.dtsi      | 157 ++++++++++++++++++
 arch/arm64/boot/dts/marvell/armada-3900.dtsi  |  55 ++++++
 5 files changed, 419 insertions(+)
 create mode 100644 arch/arm64/boot/dts/marvell/armada-3900-db-vd-A.dts
 create mode 100644 arch/arm64/boot/dts/marvell/armada-3900-db-vd-B.dts
 create mode 100644 arch/arm64/boot/dts/marvell/armada-3900-db.dtsi
 create mode 100644 arch/arm64/boot/dts/marvell/armada-3900.dtsi

diff --git a/arch/arm64/boot/dts/marvell/Makefile b/arch/arm64/boot/dts/marvell/Makefile
index 4bcbc356e557..4f0d459aa21d 100644
--- a/arch/arm64/boot/dts/marvell/Makefile
+++ b/arch/arm64/boot/dts/marvell/Makefile
@@ -24,3 +24,5 @@ dtb-$(CONFIG_ARCH_MVEBU) += armada-8160-pd.dtb
 dtb-$(CONFIG_ARCH_MVEBU) += armada-8162-db.dtb
 dtb-$(CONFIG_ARCH_MVEBU) += armada-8164-db.dtb
 dtb-$(CONFIG_ARCH_MVEBU) += armada-8320-pd.dtb
+dtb-$(CONFIG_ARCH_MVEBU) += armada-3900-db-vd-A.dtb
+dtb-$(CONFIG_ARCH_MVEBU) += armada-3900-db-vd-B.dtb
diff --git a/arch/arm64/boot/dts/marvell/armada-3900-db-vd-A.dts b/arch/arm64/boot/dts/marvell/armada-3900-db-vd-A.dts
new file mode 100644
index 000000000000..65bdc42573c8
--- /dev/null
+++ b/arch/arm64/boot/dts/marvell/armada-3900-db-vd-A.dts
@@ -0,0 +1,102 @@
+/*
+ * Copyright (C) 2018 Marvell Technology Group Ltd.
+ *
+ * This file is dual-licensed: you can use it either under the terms
+ * of the GPLv2 or the X11 license, at your option. Note that this dual
+ * licensing only applies to this file, and not this project as a
+ * whole.
+ *
+ *  a) This library is free software; you can redistribute it and/or
+ *     modify it under the terms of the GNU General Public License as
+ *     published by the Free Software Foundation; either version 2 of the
+ *     License, or (at your option) any later version.
+ *
+ *     This library is distributed in the hope that it will be useful,
+ *     but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *     GNU General Public License for more details.
+ *
+ * Or, alternatively,
+ *
+ *  b) Permission is hereby granted, free of charge, to any person
+ *     obtaining a copy of this software and associated documentation
+ *     files (the "Software"), to deal in the Software without
+ *     restriction, including without limitation the rights to use,
+ *     copy, modify, merge, publish, distribute, sublicense, and/or
+ *     sell copies of the Software, and to permit persons to whom the
+ *     Software is furnished to do so, subject to the following
+ *     conditions:
+ *
+ *     The above copyright notice and this permission notice shall be
+ *     included in all copies or substantial portions of the Software.
+ *
+ *     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ *     EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ *     OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ *     NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ *     HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ *     WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ *     FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ *     OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/*
+ * Device Tree file for Marvell Armada 3900 validation development board
+ * This board file supports "A" configuration of the board
+ */
+
+#include <dt-bindings/gpio/gpio.h>
+#include "armada-3900-db.dtsi"
+
+/ {
+	model = "Marvell Armada 3900 Validation Development board setup(A)";
+	compatible = "marvell,armada3900-db-vd",
+		     "marvell,armada-ap807-quad", "marvell,armada-ap807";
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+};
+
+&uart0 {
+	status = "okay";
+};
+
+&cp0_i2c0 {
+	status = "okay";
+};
+
+&cp0_spi1 {
+	status = "okay";
+};
+
+&cp0_sdhci0 {
+	status = "okay";
+};
+
+&spi0 {
+	status = "okay";
+};
+
+&cp0_nand_controller {
+	status = "okay";
+};
+
+&cp0_usb3_0 {
+	status = "okay";
+};
+
+&cp0_usb3_1 {
+	status = "disabled";
+};
+
+&cp0_pcie0 {
+	status = "okay";
+};
+
+&cp0_pcie2 {
+	status = "okay";
+};
+
+
+
diff --git a/arch/arm64/boot/dts/marvell/armada-3900-db-vd-B.dts b/arch/arm64/boot/dts/marvell/armada-3900-db-vd-B.dts
new file mode 100644
index 000000000000..669df36c08ef
--- /dev/null
+++ b/arch/arm64/boot/dts/marvell/armada-3900-db-vd-B.dts
@@ -0,0 +1,103 @@
+/*
+ * Copyright (C) 2018 Marvell Technology Group Ltd.
+ *
+ * This file is dual-licensed: you can use it either under the terms
+ * of the GPLv2 or the X11 license, at your option. Note that this dual
+ * licensing only applies to this file, and not this project as a
+ * whole.
+ *
+ *  a) This library is free software; you can redistribute it and/or
+ *     modify it under the terms of the GNU General Public License as
+ *     published by the Free Software Foundation; either version 2 of the
+ *     License, or (at your option) any later version.
+ *
+ *     This library is distributed in the hope that it will be useful,
+ *     but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *     GNU General Public License for more details.
+ *
+ * Or, alternatively,
+ *
+ *  b) Permission is hereby granted, free of charge, to any person
+ *     obtaining a copy of this software and associated documentation
+ *     files (the "Software"), to deal in the Software without
+ *     restriction, including without limitation the rights to use,
+ *     copy, modify, merge, publish, distribute, sublicense, and/or
+ *     sell copies of the Software, and to permit persons to whom the
+ *     Software is furnished to do so, subject to the following
+ *     conditions:
+ *
+ *     The above copyright notice and this permission notice shall be
+ *     included in all copies or substantial portions of the Software.
+ *
+ *     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ *     EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ *     OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ *     NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ *     HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ *     WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ *     FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ *     OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/*
+ * Device Tree file for Marvell Armada 3900 validation development board
+ * This board file supports "B" configuration of the board
+ */
+
+#include <dt-bindings/gpio/gpio.h>
+#include "armada-3900-db.dtsi"
+
+/ {
+	model = "Marvell Armada 3900 Validation Development board setup(B)";
+	compatible = "marvell,armada3900-db-vd",
+		     "marvell,armada-ap807-quad", "marvell,armada-ap807";
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+
+};
+
+&uart0 {
+	status = "okay";
+};
+
+&cp0_i2c0 {
+	status = "okay";
+};
+
+&cp0_spi1 {
+	status = "okay";
+};
+
+&cp0_sdhci0 {
+	status = "okay";
+};
+
+&spi0 {
+	status = "okay";
+};
+
+&cp0_nand_controller {
+	status = "okay";
+};
+
+&cp0_usb3_0 {
+	status = "okay";
+};
+
+&cp0_usb3_1 {
+	status = "disabled";
+};
+
+&cp0_pcie0 {
+	status = "okay";
+};
+
+&cp0_pcie2 {
+	status = "okay";
+};
+
+
+
diff --git a/arch/arm64/boot/dts/marvell/armada-3900-db.dtsi b/arch/arm64/boot/dts/marvell/armada-3900-db.dtsi
new file mode 100644
index 000000000000..e07062b0f98a
--- /dev/null
+++ b/arch/arm64/boot/dts/marvell/armada-3900-db.dtsi
@@ -0,0 +1,157 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Device Tree file for Marvell Armada 7040 development board platform
+ */
+
+#include <dt-bindings/gpio/gpio.h>
+#include "armada-3900.dtsi"
+
+/ {
+	model = "Marvell Armada 3900";
+	compatible = "marvell,armada3900", "marvell,armada-ap807-quad",
+		     "marvell,armada-ap807";
+	#address-cells = <2>;
+	#size-cells = <2>;
+
+	memory@0 {
+		device_type = "memory";
+		reg = <0x0 0x0 0x0 0x80000000>;
+	};
+
+	/* USB3 VBus controller through GPIO */
+	cp0_usb3_0_phy: cp0-usb3-0-phy {
+		compatible = "usb-nop-xceiv";
+		vcc-supply = <&cp0_reg_usb3_0_vbus>;
+	};
+	cp0_usb3_1_phy: cp0-usb3-1-phy {
+		compatible = "usb-nop-xceiv";
+		vcc-supply = <&cp0_reg_usb3_1_vbus>;
+	};
+	cp0_reg_usb3_0_vbus: cp0_usb3_vbus@0 {
+		compatible = "regulator-fixed";
+		gpio = <&cp0_gpio1 12 GPIO_ACTIVE_HIGH>;
+		regulator-name = "cp0-xhci0-vbus";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		enable-active-high;
+		regulator-always-on;
+	};
+	cp0_reg_usb3_1_vbus: cp0_usb3_vbus@1 {
+		compatible = "regulator-fixed";
+		gpio = <&cp0_gpio1 13 GPIO_ACTIVE_HIGH>;
+		regulator-name = "cp0-xhci1-vbus";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		enable-active-high;
+		regulator-always-on;
+	};
+};
+
+&uart0 {
+	pinctrl-0 = <&uart0_pins>;
+	pinctrl-names = "default";
+};
+
+&spi0 {
+	spi-flash@0 {
+		#address-cells = <0x1>;
+		#size-cells = <0x1>;
+		compatible = "jedec,spi-nor";
+		reg = <0x0>;
+		spi-max-frequency = <50000000>;
+		partition@0 {
+			label = "boot";
+			reg = <0x0 0x500000>;
+		};
+		partition@500000 {
+			label = "env";
+			reg = <0x500000 0xb00000>;
+		};
+	};
+};
+
+&cp0_config_space {
+	ranges = /* internal regs */
+		 <0x0 0x0 0xf2000000 0x2000000>,
+		 /* SPI1-DEV0 */
+		 <0x2000000 0 0xf9000000 0x1000000>;
+};
+
+&cp0_i2c0 {
+	status = "disabled";
+	clock-frequency = <100000>;
+
+	eeprom@57 {
+		compatible = "atmel,24c64";
+		reg = <0x57>;
+		pagesize = <32>;
+	};
+
+	eeprom@50 {
+		compatible = "atmel,24c64";
+		reg = <0x50>;
+		pagesize = <32>;
+	};
+};
+
+&cp0_nand_controller {
+	nand@0 {
+		reg = <0>;
+		label = "main-storage";
+		nand-rb = <0>;
+		nand-ecc-mode = "hw";
+		marvell,nand-keep-config;
+		nand-on-flash-bbt;
+		nand-ecc-strength = <8>;
+		nand-ecc-step-size = <512>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "U-Boot";
+				reg = <0 0x400000>;
+			};
+			partition@1000000 {
+				label = "rootfs";
+				reg = <0x1000000 0x1f800000>;
+			};
+			partition@20800000 {
+				label = "alt_rootfs";
+				reg = <0x20800000 0x1f800000>;
+			};
+		};
+	};
+};
+
+&cp0_usb3_0 {
+	status = "disabled";
+	usb-phy = <&cp0_usb3_0_phy>;
+	phys = <&cp0_comphy1 0>;
+	phy-names = "usb";
+};
+
+&cp0_usb3_1 {
+	status = "disabled";
+	usb-phy = <&cp0_usb3_1_phy>;
+	phys = <&cp0_comphy4 1>;
+	phy-names = "usb";
+};
+
+&cp0_pcie0 {
+	ranges = <0x81000000 0 0xf7000000 0  0xf7000000 0 0x10000
+		  0x82000000 0 0xc0000000 0  0xc0000000 0 0x20000000>;
+	phys = <&cp0_comphy0 0>;
+	phy-names = "pcie-phy0";
+};
+
+&cp0_pcie2 {
+	ranges = <0x81000000 0 0xf7020000 0  0xf7020000 0 0x10000
+		  0x82000000 0 0xf8000000 0  0xf8000000 0 0xf00000>;
+	phys = <&cp0_comphy5 2>;
+	phy-names = "pcie-phy2";
+};
+
+
diff --git a/arch/arm64/boot/dts/marvell/armada-3900.dtsi b/arch/arm64/boot/dts/marvell/armada-3900.dtsi
new file mode 100644
index 000000000000..8383a1ccb64e
--- /dev/null
+++ b/arch/arm64/boot/dts/marvell/armada-3900.dtsi
@@ -0,0 +1,55 @@
+/*
+ * Copyright (C) 2018 Marvell Technology Group Ltd.
+ *
+ * This file is dual-licensed: you can use it either under the terms
+ * of the GPLv2 or the X11 license, at your option. Note that this dual
+ * licensing only applies to this file, and not this project as a
+ * whole.
+ *
+ *  a) This library is free software; you can redistribute it and/or
+ *     modify it under the terms of the GNU General Public License as
+ *     published by the Free Software Foundation; either version 2 of the
+ *     License, or (at your option) any later version.
+ *
+ *     This library is distributed in the hope that it will be useful,
+ *     but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *     GNU General Public License for more details.
+ *
+ * Or, alternatively,
+ *
+ *  b) Permission is hereby granted, free of charge, to any person
+ *     obtaining a copy of this software and associated documentation
+ *     files (the "Software"), to deal in the Software without
+ *     restriction, including without limitation the rights to use,
+ *     copy, modify, merge, publish, distribute, sublicense, and/or
+ *     sell copies of the Software, and to permit persons to whom the
+ *     Software is furnished to do so, subject to the following
+ *     conditions:
+ *
+ *     The above copyright notice and this permission notice shall be
+ *     included in all copies or substantial portions of the Software.
+ *
+ *     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ *     EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ *     OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ *     NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ *     HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ *     WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ *     FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ *     OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/*
+ * Device Tree file for the Armada 3900 SoC, made of an AP807 Quad and
+ * one CP110.
+ */
+
+#include "armada-ap807-quad.dtsi"
+#include "armada-70x0.dtsi"
+
+/ {
+	model = "Marvell Armada 3900";
+	compatible = "marvell,armada3900", "marvell,armada-ap807-quad",
+		     "marvell,armada-ap807";
+};
-- 
2.17.1

